sql/item_strfunc.cc

```diff
@@ -2947,6 +2947,10 @@ bool Item_func_soundex::fix_length_and_dec(THD *thd)
   if (agg_arg_charsets_for_string_result(collation, args, 1))
     return TRUE;
   DBUG_ASSERT(collation.collation != NULL);
+  if (collation.collation->mbmaxlen > 1)
```

Comment:
Feel free to ignore this: SOUNDEX() should never produce anything but ASCII codes. I question the usefulness of preserving the first argument's collation as a result collation. 
If we operate under the above assumption the whole algorithm will also be simpler: convert to ascii and then do the SOUNDEX on the ascii itself.